Logs Integrations Contributor Doc Updates (#24411)
DataDog/documentation: content/en/developers/integrations/log_pipeline.md

  tag: Documentation
  text: Learn about log pipelines
title: Create a Log Pipeline
description: Learn how to create a Datadog Log Integration.
---
## Overview

This page walks Technology Partners through creating a log pipeline. A log pipeline is required if your integration sends logs to Datadog.

## Integration prerequisites

In developing your integration to send logs to Datadog follow these guidlines to ensure the best experience for your users.

### Guidelines

> Before creating a log pipeline, consider the following guidlines and practices:

Reviewer: ```suggestion
Before creating a log pipeline, consider the following guidelines and best practices:
```

## Development process
In developing your integration to send logs to Datadog follow these guidlines to ensure the best experience for your users.

### Guidelines

When creating a log pipeline, consider the following best practices:

Map your data to Datadog's standard attributes
: Centralizing logs from various technologies and applications can generate tens or hundreds of different attributes in a Log Management environment. Integrations must rely as much as possible on the [standard naming convention][17].
Before creating a log pipeline, consider the following guidlines and practices:

Set the `source` tag to the integration name.

> : Datadog recommends that the `source` tag is set to `<integration_name>` and that the `service` tag is set to the name of the service that produces the telemetry. For example, the `service` tag can be used to differentiate logs by product line. </br></br> For cases where there aren't different services, set `service` to the same value as `source`. The `source` and `service` tags must be non-editable by the user because the tags are used to enable integration pipelines and dashboards. The tags can be set in the payload or through the query parameter, for example, `?ddsource=example&service=example`. </br></br> The `source` and `service` tags must be in lowercase. 

Reviewer: We should keep this section in the doc. Questions have come up in the past on source & service as well as ways to send the tags to DD.

Author: Hey @emarsha94 I actually have kept this but it appears lower in the text.
The problem I have is originally the doc refers to "Datadog Integrations" then Log Pipelines, so the concepts of Service, Integration Name and Source are confused with the actual Integration. Rather, I put the information about `source` and `service` at lines 85 onwards.

Author: Done - added this in a new commit.

> : Tags can be set as key-value attributes in the JSON body of your integration's logs payload. Datadog recommends allowing users to set custom tags for an integration. If the integation uses the [Send Logs API documentation][1] tags can optionally be set using the `ddtags=<TAGS>` query parameter.

Reviewer: ```suggestion
: Tags can be set as key-value attributes in the JSON body of your integration's logs payload. Datadog recommends allowing users to set custom tags for an integration. If the integration [sends logs via the API][1], tags can optionally be set using the `ddtags=<TAGS>` query parameter.
```

* Facets provide a straightforward interface for filtering logs. They are leveraged in Log Management autocomplete features, allowing users to find and aggregate key information found in their logs.
* Facets allow for attributes with low readability to be renamed with a label that is easier to understand. For example: `@deviceCPUper` → `Device CPU Utilization Percentage`.

You can create [facets][12] in the [Log Explorer][16].

**Add a facet or measure**:

<div class="alert alert-info">Quantitative facets, called "Measures", allow users to filter logs over a range of numeric values using relational operators.  
For example, a measure for a latency attribute allows users to search for all logs greater-than a certain duration. </div>

1. Click on a log that contains the attribute you want to add a facet or measure for. 
2. In the log panel, click the Cog icon next to the attribute.
3. Select **Create facet/measure for @attribute**.
4. For a measure, to define the unit, click **Advanced options**. Select the unit based on what the attribute represents.
**Note**: Define the [unit][11] of a measure based on what the attribute represents.
5. Click **Add**.

> To help navigate the facet list, facets are grouped together. For fields specific to the integration logs, create a **single group with the same name** as the `source` tag. 

Reviewer: Can we keep this? All log integrations should have their own group for their custom facets

Reviewer: I see that this is mentioned below. I think we might need to move this? If a partner creates a facet based on the instructions here then they won't know to add this to the integration group unless they keep reading. I'd either keep the group instructions here or move this after requirements? For example :

1. Facet overview
2. Requirements
3. Create facet instructions
4. Configure and edit

Author: I think visually, the requirements read-out can be a bit challenging initially - getting them to do the basic creation of the facet then have it fit the requirements might be easier.

I can add a quick "Facet group" instruction here and point the user down the docs for additional requirements on that.
So basically have this be step 5.
> To help navigate the facet list, facets can be grouped together. If a Group does not already exist for your Facets please read For fields specific to the integration logs, create a **single group with the same name** as the `source` tag.

I'll recommit and we can review if this is needed then.

Author: I've added this in a new commit.

Log integrations consist of two sets of assets: [pipelines][13] and associated [facets][12]. Centralizing logs from various technologies and applications can produce many unique attributes. To use out-of-the-box dashboards, Technology Partner Integrations should rely on Datadog's [standard naming convention][17] when creating integrations.

After finalizing your Datadog Integration design and successfully sending logs to Datadog's log endpoint(s), define your log pipelines and facets to enrich and structure your integration's Logs.

For information about becoming a Datadog Technology Partner, and gaining access to an integration development sandbox, read [Build an Integration][18].

<div class="alert alert-warning">To be reviewed by Datadog's integration team, log integrations must include assets and have pipeline processors or facets.</div>

### Pipelines overview

Logs sent to Datadog are processed in [log pipelines][13] using pipeline processors. These processors allow users to parse, remap, and extract attribute information, enriching and standardizing logs for use across the platform.

#### Create a pipeline

> Create a log pipeline to process specified logs with pipeline processors. Once you've set-up a pipeline add processors to enrich and structure the logs further.

Reviewer 2: Moved this sentence to the end of this section.

```suggestion
Create a log pipeline to process specified logs with pipeline processors.
```

1. From the [**Pipelines**][3] page, click **+ New Pipeline**.
2. In the **Filter** field, enter the unique `source` tag that defines the log source for the Technology Partner's logs. For example, `source:okta` for the Okta integration. 
**Note**: Make sure that logs sent through the integration are tagged with the correct source tags before they are sent to Datadog.
3. Optionally, add tags and a description.
4. Click **Create**.

> 

Reviewer 2: ```suggestion

After you set-up a pipeline, add processors to enrich and structure the logs further.
```

Author: Makes sense ! thanks for catching

#### Add pipeline processors

Before defining your pipeline processors, review [Datadog's Standard Attributes][6].

> Use processors within your pipelines to enrich, restructure your data, and generate log attributes. For a list of all log processors, see the [Processors][10] documentation.

Reviewer 2: ```suggestion
Use processors within your pipelines to enrich and restructure your data, and generate log attributes. For a list of all log processors, see the [Processors][10] documentation.
```

  tag: Documentation
  text: Learn about log pipelines
title: Create a Log Pipeline
description: Learn how to create a Datadog Log Integration Pipeline.
---
## Overview

This page walks Technology Partners through creating a log pipeline. A log pipeline is required if your integration sends logs to Datadog.

## Integration prerequisites

> In developing your integration to send logs to Datadog follow these guidlines to ensure the best experience for your users.

Reviewer 3: guidelines*

Would suggest:

When developing your integration to send logs to Datadog, follow these guidelines to ensure the best experience for your users.

Set the integration's logs `source` tag to the integration name
: Datadog recommends that the `source` tag is set to `<integration_name>` for an application. `source` must be set before sending logs to Datadog's endpoints as it cannot be remapped in the Datadog UI. </br></br> The `source` tag must be in lowercase and must not be editable by users as it is used to enable integration pipelines and dashboards.

Avoid sending logs that contain arrays in the JSON body whenever possible 
: While it's possible to send array data in your logs, Datadog recommends avoiding arrays as they cannot be [faceted][24].

Do not log Datadog API Keys
: Datadog API Keys can either be passed in the Header or as part of the HTTP Path of your API requests. See [Send Logs API documentation][1] for examples. Avoid logging the API Key in your setup.

Do not use Datadog Application Keys
: Datadog Application Keys are not required to send logs using the HTTP endpoint. 

## Create log integration assets

You can create and design your log integration assets directly within your Datadog partner account.

> Log integrations are made of two sets of assets, [pipelines][13] and accompanying [facets][12]. Centralizing logs from various technologies and applications can generate large amounts of unique attributes. To take advantage of out-of-the-box dashboards, Technology Partner Integrations should rely on Datadog’s [standard naming convention][17] when creating integrations.

Reviewer 3: "associated facets" to simplify sentence

**Set-up a log pipeline**

1. From the [**Pipelines**][3] page, click **+ New Pipeline**.
2. In the **Filter** field, enter the unique `source` tag that defines the log source for the Technology Partner's logs. For example, `source:okta` for the Okta integration. 
**Note**: Make sure that logs sent through the integration are tagged with the correct source tags before they are sent to Datadog.
3. Optionally, add tags and a description.
4. Click **Create**.

#### Add pipeline processors

Before defining your pipeline processors make sure to review [Datadog's Standard Attributes][6].

Use processors within your pipelines to enrich, restructure your data, and generate log attributes. For a list of all log processors, see [Processors][10].

##### Requirements

Map the application's logs attributes to [Datadog's Standard Attributes][6]
: Use the [Attribute Remapper][5] to map attribute keys to [Datadog Standard Attributes][6] where possible. For example, an attribute for a network service client IP value should be remapped to `network.client.ip`.

Map the log `service` tag to the name of the [service][26] producing telemetry

> : Use the [Service Remapper][7] to remap the `service` attribute. When the source and service are not different services, set `service` to the same value as the `source` tag. `service` tags must be in lowercase. 

Reviewer 3: is it valid use case to have same source and service value?

If yes, could we rephrase to say "When the source and service share the same value, set..."

Reviewer 3: Why is it "service tags" and not "service tag"? we can have different service?

> Map the log's internal timestamp to it's official Datadog timestamp

Reviewer 3: "its official" and not "it's"

> : Use the [Date Remapper][4] to define the official timestamp for logs. If the log timestamp does not already map to a [standard date attribute][28] it defaults to the time of ingestion at Datadog's endpoint.

Author: @estherk15 I've accepted most of the suggestion you've added thanks ! Can you give this one a look at for clarity and let me know your thoughts ?
Otherwise all of it LGTM !

Thanks !

> : Use the [Date Remapper][4] to define the official timestamp for logs. If the log timestamp does not already map to a [standard date attribute][28] it defaults to the time of ingestion at Datadog's endpoint.

Reviewer 3: not clear for me - if it defaults to the time of ingestion and then I remap, will it override this?

Map the custom status attributes of the logs to the official Datadog `status` attribute
: Use a [Status Remapper][25] to remap the `status` of a log, or a [Category Processor][19] for statuses mapped to a range (as with HTTP status codes).

Map the custom message attribute of the logs to the official Datadog `message` attribute
: Use the [message remapper][9] to define the official message of the log if application logs do not map to the standard message attribute. This allows users to search for logs using free text.

Set a namespace for custom attributes within your logs

> : Generic log attributes that do not map to a [Datadog Standard Attribute][6] must be namespaced if they are mapped to [Facets][14].

Reviewer 3: can we add example?

Author: Examples are available in a few places.

Use the [Attribute Remapper][5] to set attribute keys to a new namespaced attribute. 

1. Expand the newly created pipeline and click the "Add Processor" button to begin building your pipeline using processors.

> 2. If the integration's logs aren't `JSON` formatted, extract attribute information using the [Grok Processor][8]. Use grok processors to parse out attributes and enrich logs prior to remapping or further processing.

Reviewer 3: can we flag this one as optional?

Author: I think it's best we keep that as a direct instruction due to how frequently people might skip optional instructions.

> 3. After extracting log attributes, remap them to [Datadog's Standard Attributes][6] where possible using [Attribute Remappers][5].

Reviewer 3: and then say "Once log are "Json" formatted...."

4. Set the timestamp of an integration's logs to be its official Datadog timestamp using the [Date Remapper][4].
5. For more advanced processing and data transformations, make use of additional [processors][10].  
For example, the `Arithmetic Processor` can be used to calculate information based off of attributes, or the `String Builder Processor` can concatenate multiple string attributes. 

**Tips**
* Remove original attributes when remapping log attributes by using `preserveSource:false`. This helps avoid confusion and removes duplicates.
* To maintain optimal grok parsing performance, avoid wildcard matchers such as `%{data:}` and `%{regex(".*"):}`. Make your parsing statements as specific as possible.
* Take the free course [Going Deeper with Logs Processing][20] for an overview on writing processors and leveraging standard attributes. 

### Facets overview

Facets are specific qualitative or quantitative attributes that can be used to filter and narrow down search results. While facets are not strictly necessary for filtering search results, they play a crucial role in helping users understand the available dimensions for refining their search. 

> Facets for standard attributes are automatically added by Datadog when a pipeline is published. Review if the attribute should be remapped to a [Datadog Standard Attribute][6] instead. 

Reviewer 3: instead of what?

Not all attributes are meant to be used as a facet. The need for facets in integrations is focused on two things:
* Facets provide a straightforward interface for filtering logs. They are leveraged in Log Management autocomplete features, allowing users to find and aggregate key information found in their logs.
* Facets allow for attributes with low readability to be renamed with a label that is easier to understand. For example: `@deviceCPUper` → `Device CPU Utilization Percentage`.

> You can create [facets][12] in the [Log Explorer][16].

Reviewer 3: should we mention that logs need to be indexed to show in log explorer?

Author: Added the word indexed in the first paragraph. Should resolve this.

Custom facets must not duplicate an existing Datadog Facet
: To avoid confusion with existing out-of-the-box Datadog facets, avoid creating custom facets that duplicate an existing facets mapped to [Datadog Standard Attributes][6].

Custom facets must be grouped under the `source` name
: When creating a custom facet a group should be assigned. Set the `Group` value to the `source`, same as the integration's name.

Custom facets must have the same data type as the mapped attribute
: Set the facet data type (String, Boolean, Double, or Integer) to the same type as the Attribute mapped to it. Mismatched types prevent the facet from being used as intended and can cause it to populate incorrectly.

**Add a facet or measure**

1. Click on a log that contains the attribute you want to add a facet or measure for. 
2. In the log panel, click the Cog icon next to the attribute.
3. Select **Create facet/measure for @attribute**.
4. For a measure, to define the unit, click **Advanced options**. Select the unit based on what the attribute represents.
**Note**: Define the [unit][11] of a measure based on what the attribute represents.

> 5. To help navigate the facet list specify a facet **Group**. If one does not exist enter the name of the group matching the source tag and a description of the new group, and select **New group**.

Reviewer 3: add a comma after list or change sentence to "Specify a facet group to help navigate the facet list"

Reviewer 3: and then "If the facet group does not exist, enter the name"...

6. Click **Add** to create the facet.

#### Configure and edit facets

1. In the log panel, click the Cog icon next to the attribute that you want to configure or group.
2. Select **Edit facet/measure for @attribute**. If there isn't a facet for the attribute yet, select **Create facet/measure for @attribute**.
3. Click **Add** or **Update** when done.

**Tips**

> * Measures should have a unit where possible. Measures can be assigned a [unit][27]. Two families of units are available, `TIME` and `BYTES`, with units such as `millisecond` or `gibibyte`.

Reviewer 3: should we tell to add processors to convert to these units if current data is in seconds or hours for example?

Author: There's a link to the document with the full set of units.

* Facets can be assigned a description. A clear description of the facet can help users understand how to best use it.

> * If you redefine an attribute and keep both, define a facet on a single one.

Reviewer 3: If you remap*

> * When manually configuring facets in `.yaml` files, note they are assigned a `source`. This refers to where the attribute is captured from and can be `log` for attributes or `tag` for tags.

Reviewer 3: This mentions yaml files even if it was never mentioned in the page until now, quite confusing

Author: added some more descriptive wording.

{{< img src="developers/integrations/export_pipeline.png" alt="Click the Export Pipeline icon to export your log pipeline in Datadog" width="50%">}}

> Include sample raw logs with **all** the attributes you expect to be sent into Datadog by your integration. Raw logs comprise of the raw messages generated directly from the source application **before** they have are sent to Datadog.

Reviewer: ```suggestion
Include sample raw logs with **all** the attributes you expect to be sent into Datadog by your integration. Raw logs comprise of the raw messages generated directly from the source application **before** they are sent to Datadog.
```

{{< img src="developers/integrations/export_pipeline.png" alt="Click the Export Pipeline icon to export your log pipeline in Datadog" width="50%">}}

Include sample raw logs with **all** the attributes you expect to be sent into Datadog by your integration. Raw logs comprise of the raw messages generated directly from the source application **before** they are sent to Datadog.

Exporting your log pipeline includes two YAML files:

* One with the log pipeline, which includes custom facets, attribute remappers, and grok parsers. The exported file is named `pipeline-name.yaml`.

> * One with the raw example logs with an empty result. The exported file is named `pipeline-name_test.yaml`

Reviewer 3: what does it mean "empty result"?

Author: It's part of the submission process. In a next version we can add images to describe this.

2. Not providing the result of running the raw logs you provided against your pipeline. If the resulting output from the validation is accurate, take that output and add it to the `result` field in the YAML file containing the raw example logs.
3. If you send `service` as a parameter, instead of sending it in the log payload, you must include the `service` field below your log samples within the yaml file.

After validations pass, Datadog creates and deploys the new log integration assets. If you have any questions, add them as comments in your pull request. Datadog team members respond within 2-3 business days.

> ## Further reading

Reviewer 3: Should we mention how it works for updating an integration?

Author: Nope it's the same process.
